File: service/simple-js/seller.js
var autobahn = require('autobahn');
var xbr = require('autobahn-xbr');
var web3 = require('web3');
var BN = web3.utils.BN;
var uuid = require('uuid-buffer');

console.log('Running on Autobahn ' + autobahn.version);

// WAMP connection
var connection = new autobahn.Connection({
    realm: "realm1",
    transports: [
        {
            //url: 'wss://continental2.crossbario.com/ws',
            url: 'ws://localhost:8080/ws',
            type: 'websocket',
            serializers: [ new autobahn.serializer.CBORSerializer() ],
        }
    ]
});

// callback fired upon new WAMP session
connection.onopen = function (session, details) {

    console.log("WAMP session connected:", details);

    // ethereum private key of buyer1 delegate1
    const seller1_delegate2_pkey = "0xa453611d9419d0e56f499079478fd72c37b251a94bfde4d19872c44cf65386e3";

    // market maker ethereum address
    const marketmaker_adr = "0x3E5e9111Ae8eB78Fe1CC3bb8915d5D461F3Ef9A9";

    // the XBR token has 18 decimals
    const decimals = new BN('1000000000000000000');

    // price in XBR per key
    const price = new BN(35).mul(decimals);

    // key rotation interval in seconds
    const key_rotation_interval = 10;

    // API ID of the interface of the offered service
    const apiID = uuid.toBuffer('bd65e220-aef7-43c0-a801-f8d95fa71f39');

    // instantiate a simple seller
    var seller = new xbr.SimpleSeller(marketmaker_adr, seller1_delegate2_pkey);
    var counter = 1;

    seller.add(apiID, 'io.crossbar.example', price, key_rotation_interval);

    var do_publish = function() {
        const payload = {"data": "js-seller", "counter": counter};

        // encrypt the XBR payload, potentially automatically rotating & offering a new data encryption key
        let [key_id, enc_ser, ciphertext] = seller.wrap(apiID, 'io.crossbar.example', payload);

        const options = {acknowledge: true};

        session.publish("io.crossbar.example",
                        [key_id, enc_ser, ciphertext],
                        {},
                        options).then(
            function (pub) {
                console.log("Published event " + pub.id);
            },
            function (err) {
                console.log("Failed to publish event", err);
            }
        );

        counter += 1;
        setTimeout(do_publish, 1000);
    };

    // start selling
    seller.start(session).then(
        // success: we've got an active paying channel with remaining balance ..
        function (balance) {
            console.log("Seller has started, remaining balance in active paying channel is " + balance.div(decimals) + " XBR");
            do_publish();
        },
        // we don't have an active paying channel => bail out
        function (error) {
            console.log("Failed to start seller:", error);
            process.exit(1);
        }
    );
};

// open WAMP session
connection.open();

File: service/simple-python/seller.py
#!/usr/bin/env python3
"""Simple Seller example.
"""

import binascii
import os
from time import sleep
from uuid import UUID

import txaio
txaio.use_twisted()

from autobahn.wamp.types import PublishOptions
from autobahn.twisted.component import Component, run
from autobahn.twisted.util import sleep
from autobahn.twisted.xbr import SimpleSeller

DEFAULT_AUTHID = 'team4'
DEFAULT_TICKET = 'knife23 fork spoon cutlery'

TOPIC = 'io.crossbar.example'
SERVICE_NAME = 'simple seller python example service'
SERVICE_TYPE = 'seller'

comp = Component(
    transports=os.environ.get('XBR_INSTANCE', 'wss://continental2.crossbario.com/ws'),
    realm=os.environ.get('XBR_REALM', 'realm1'),
    authentication={
        'ticket': {
            'authid': os.environ.get('XBR_AUTHID', DEFAULT_AUTHID),
            'ticket': os.environ.get('XBR_TICKET', DEFAULT_TICKET),
            'authextra': {
                'service_name': SERVICE_NAME,
                'type': SERVICE_TYPE,
            },
        },
    },
    extra={
        'market_maker_adr': os.environ.get('XBR_MARKET_MAKER_ADR', '0x3e5e9111ae8eb78fe1cc3bb8915d5d461f3ef9a9'),
        'seller_privkey': os.environ.get('XBR_SELLER_PRIVKEY', '0xe106047103208a20498ccf40079e3c3a37b7f560fd57d7edaf9e609ab9704e0a'),
    }
)

running = False


@comp.on_join
async def joined(session, details):
    print('Seller session joined', details)
    global running
    running = True

    market_maker_adr = binascii.a2b_hex(session.config.extra['market_maker_adr'][2:])
    print('Using market maker adr:', session.config.extra['market_maker_adr'])

    seller_privkey = binascii.a2b_hex(session.config.extra['seller_privkey'][2:])

    api_id = UUID('627f1b5c-58c2-43b1-8422-a34f7d3f5a04').bytes
    topic = TOPIC
    counter = 1

    seller = SimpleSeller(market_maker_adr, seller_privkey)
    price = 35 * 10 ** 18
    interval = 10
    seller.add(api_id, topic, price, interval, None)

    balance = await seller.start(session)
    balance = int(balance / 10 ** 18)
    print("Remaining balance: {} XBR".format(balance))

    while running:
        payload = {'data': 'py-seller', 'counter': counter}
        key_id, enc_ser, ciphertext = await seller.wrap(api_id,
                                                        topic,
                                                        payload)
        pub = await session.publish(topic, key_id, enc_ser, ciphertext,
                                    options=PublishOptions(acknowledge=True))

        print('Published event {}: {}'.format(pub.id, payload))

        counter += 1
        await sleep(1)


@comp.on_leave
def left(session, details):
    print('Seller session left', details)
    global running
    running = False


if __name__ == '__main__':
    run([comp])
